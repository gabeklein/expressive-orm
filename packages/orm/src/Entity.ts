import { Field, init, num, underscore } from './fields';
import { desc, equal, Where } from './where';

type Values<T extends Type> = {
  [K in Exclude<keyof T, keyof Type> as (
    T[K] extends Type.Class ? never : K
  )]: T[K];
}

type Compat<T extends Type> = Partial<Values<T>>;

type Insert<T extends Type> = {
  [K in keyof Omit<T, keyof Type> as (
    undefined extends T[K] ? never : (T[K] extends Type.Class<any> ? never : K)
  )]: T[K];
} & {
  [K in keyof Omit<T, keyof Type> as (
    undefined extends T[K] ? (T[K] extends Type.Class<any> ? never : K) : never
  )]?: T[K];
};

declare namespace Type {
  type Class<T extends Type = Type> = (abstract new (...args: any[]) => T) & typeof Type;

  type Instance<T> = T extends Class<infer U> ? U : T;

  type KeyOf<T> = T extends Class<infer U> ? keyof U : keyof T;

  type Result<T extends Type> = PromiseLike<T[]>;

  type Conds<T> = Where | T | Conds<T>[]

  type Query<T extends {}> = {
    [K in keyof T]?: Conds<T[K]>;
  };

  type Constraint = [
    field: string,
    op: string,
    value?: any
  ]

  interface Connection {
    get: (table: string, constraints: Constraint[], limit?: number, offset?: number) => Promise<any[]>;
    insert: (table: string, data: Record<string, any>, returns?: boolean) => Promise<any>;
    update: (table: string, id: string | number, data: Record<string, any>) => Promise<void>;
    remove: (table: string, id: string | number) => Promise<void>;
  }

  export {
    Class,
    Compat,
    Connection,
    Constraint,
    Insert,
    Instance,
    KeyOf,
    Query,
    Result,
    Values
  }
}

async function reject(): Promise<never> {
  throw new Error("No database connection. Please set Entity.connection to a valid Connection instance.");
}

class NotReady {
  get = reject;
  insert = reject;
  update = reject;
  remove = reject;
}

abstract class Type {
  id = num({ optional: true });

  get values() {
    const values: Record<string, any> = {};
    const type = this.constructor as typeof Type;

    for (const [key] of type.fields)
      values[key] = (this as any)[key];

    return Object.freeze(values) as Values<this>;
  };

  protected async onDelete?(): Promise<void>;

  async update(data: Type.Compat<this>) {
    if (Object.keys(data).length === 0)
      return;

    const type = this.constructor as typeof Type
    await type.connection.update(type.ref, this.id, type.into(data, true));

    for (const [key] of type.fields)
      if (key in data)
        Object.defineProperty(this, key, {
          value: (data as any)[key],
          configurable: true,
          enumerable: true,
          writable: false,
        });
  }

  async delete() {
    const { connection, ref } = this.constructor as typeof Type;

    if (typeof this.onDelete === "function")
      await this.onDelete();

    await connection.remove(ref, this.id);
  }

  static schema?: string;
  static subset?: Record<string, unknown>;
  static connection: Type.Connection = new NotReady();
  static table?: string;

  static get ref() {
    const { name, schema, table = underscore(name) } = this;
    const ref = schema ? `${schema}.${table}` : table;
    Object.defineProperty(this, 'ref', { value: ref });
    return ref;
  }

  static get fields(): Map<string, Field> {
    const fields = init(this);
    Object.defineProperty(this, 'fields', { value: fields });
    return fields;
  }

  protected static from<T extends Type>(
    this: Type.Class<T>,
    raw: Record<string, any>
  ): T {
    const entity = Object.create(this.prototype) as T;

    this.fields.forEach(field => {
      Object.assign(entity, field.input(raw));
    });

    return entity;
  }

  protected static into<T extends Type>(
    this: Type.Class<T>,
    data: Compat<T>,
    partial?: boolean
  ): Record<string, any> {
    const result: Record<string, any> = {};

    this.fields.forEach(field => {
      Object.assign(result, field.output(data, partial));
    })

    return result;
  }

  static async set<T extends Type>(this: Type.Class<T>, id: string, data: Type.Compat<T>) {
    await this.connection.update(this.ref, id, this.into(data, true));
  }

  static async new<T extends Type>(this: Type.Class<T>, data: Type.Insert<T>): Promise<T>;
  static async new<T extends Type>(this: Type.Class<T>, data: Type.Insert<T>[]): Promise<void>;
  static async new<T extends Type, U>(this: Type.Class<T>, data: Array<U>, mapFn: (item: U) => Type.Insert<T>): Promise<void>;
  static async new<T extends Type>(
    this: Type.Class<T>,
    data: Values<T> | unknown[],
    mapFn?: (item: unknown) => any
  ) {
    const insert = async (d: Values<T>, returns?: boolean) => {
      const input = this.into({ ...d as any, ...this.subset });
      const inserted = await this.connection.insert(this.ref, input, returns);
      return this.from(inserted) as T;
    };

    if (Array.isArray(data)) {
      await Promise.all(data.map(mapFn || (x => x)).map(x => insert(x)))
      return;
    }

    return insert(data, true);
  }

  static async one<T extends Type>(this: Type.Class<T>, id: number, expect?: true): Promise<T>;
  static async one<T extends Type>(this: Type.Class<T>, id: number, expect: boolean): Promise<T | undefined>;
  static async one<T extends Type>(this: Type.Class<T>, where: Type.Query<T>, expect: boolean): Promise<T | undefined>;
  static async one<T extends Type>(this: Type.Class<T>, where?: Type.Query<T>, expect?: true): Promise<T>;
  static async one<T extends Type>(this: Type.Class<T>, where?: number | Type.Query<T>, expect?: boolean): Promise<T | undefined> {
    if (typeof where === "number")
      where = { id: equal(where) };

    const [row] = await this.query(where, 1);

    if (!row && expect !== false)
      throw new Error(`No ${this.name} found with: \n${JSON.stringify(where, null, 2)}\n`);

    return row;
  }

  static async query<T extends Type>(
    from?: Type.Constraint[] | Type.Query<T>,
    limit?: number,
    offset: number = 0
  ): Promise<T[]> {
    const { connection, subset, ref } = this;

    if (!from)
      from = { id: desc() };

    if (!Array.isArray(from))
      from = buildConstraints({ ...from, ...subset });

    from = from.map(([key, op, value]) => {
      const field = this.fields.get(key);

      if(field)
        return [field.column, op, field.set(value)]
    }).filter(Boolean) as Type.Constraint[];

    const rows = await connection.get(ref, from, limit, offset);
    return rows.map(row => this.from(row) as T)
  }

  static async get<T extends Type>(
    this: Type.Class<T>,
    where?: Type.Query<T>,
    limit?: number,
    offset?: number
  ) {
    return this.query(where, limit, offset);
  }

  static async *all<T extends Type>(
    this: Type.Class<T>,
    pageSize: number,
    where?: Type.Query<T>
  ): AsyncGenerator<T[]> {
    let offset = 0;
    let done = false;

    while (!done) {
      const batch = await this.query(where, pageSize, offset);

      offset += batch.length;

      if (!batch.length) break;
      if (pageSize && batch.length < pageSize) done = true;

      yield batch;
    }
  }
}

function buildConstraints(where: Type.Query<any>) {
  const constraints: Type.Constraint[] = [];

  for (const key in where) {
    let field = where[key] as Where | Where[];

    if (!Array.isArray(field))
      field = [field];

    for (const val of field) {
      if (val === undefined || val instanceof Where && val.skip)
        continue;

      constraints.push(
        val instanceof Where
          ? [key, val.op || "=", val.value]
          : [key, '=', val]
      )
    }
  }

  return constraints;
}

export {
  Type,
  Type as default,
  Values,
  Compat,
  Insert,
};