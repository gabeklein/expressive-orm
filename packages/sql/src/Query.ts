import knex, { Knex } from 'knex';

import { Field } from './Field';
import { isTypeConstructor, Type } from './Type';

export const RelevantTable = new WeakMap<{}, Query.Table>();
declare const ENTITY: unique symbol;

declare namespace Query { 
  namespace Join {
    type Mode = "left" | "inner";

    type Where = (field: Field) => {
      is(equalTo: Field): void;
      isNot(equalTo: Field): void;
      isMore(than: Field): void;
      isLess(than: Field): void;
    }

    type Function = (on: Where) => void;

    type Object<T extends Type> = {
      [K in Type.Field<T>]?: T[K];
    }
  }

  type Assertion = symbol;

  interface Compare<T> {
    is(equalTo: T): Assertion;
    isNot(equalTo: T): Assertion;
    isMore(than: T): Assertion;
    isLess(than: T): Assertion;
  }

  interface Assert<T> extends Compare<T> {
    isAsc(): void;
    isDesc(): void;
  }

  interface Verbs <T extends Type> {
    delete(): void;
    update(values: Query.Update<T>): void;
  }

  interface Table {
    type: Type.EntityType;
    name: string | Knex.AliasDict;
    alias?: string;
  }

  type FromType<T extends Type = Type> = {
    [K in Type.Field<T>]: Exclude<T[K], null>;
  } & {
    [ENTITY]?: T
  }

  type Compat<T extends Type = Type> = {
    [K in Type.Field<T>]?: T[K];
  }

  // TODO: make this default/nullable aware.
  type Update<T extends Type> = {
    [K in Type.Field<T>]?: Exclude<T[K], undefined>;
  }

  interface Where {
    <T extends Type>(entity: Type.EntityType<T>, on?: Compat<T> | Join.Function, join?: "inner"): FromType<T>;
    <T extends Type>(entity: Type.EntityType<T>, on: Compat<T> | Join.Function, join: Join.Mode): Partial<FromType<T>>;

    <T extends Type>(field: FromType<T>): Verbs<T>;
    
    <T>(field: T): Assert<T>;
  }

  type Function<R> = (where: Where) => R;
}

interface Query<T = unknown> extends PromiseLike<T> {
  /**
   * Counts the number of rows that would be selected or modified.
   * Will ignore any side effects, such as update and delete.
   **/
  count(): Promise<number>;

  /**
   * Returns the SQL string generated by this query.
   */
  toString(): string;
}

interface SelectQuery<T = unknown> extends Query<T[]> {
  /**
   * Returns the first row that matches creteria.
   * 
   * @param orFail If true, will throw an error if no results are returned.
   */
  one(orFail?: boolean): Promise<T>;
}

/** Executable query on whole types. */
interface TypeQuery<T extends Type = any> extends SelectQuery<T> {
  update(values: Type.Insert<T>): Promise<number>;
  delete(): Promise<number>;
}

function Query<T extends Type>(from: Query.Function<T>): TypeQuery<T>;

function Query<T extends {}>(from: Query.Function<T>): SelectQuery<T>;

/**
 * Creates a new query.
 * 
 * If no selection is returned by the constructor, will return
 * the number of rows that would be selected or modified.
 */
function Query(from: Query.Function<void>): Query<number>;

function Query<T = void>(constructor: Query.Function<T>): Query | SelectQuery | TypeQuery {
  const tables = [] as Query.Table[];
  const pending = new Set<() => void>();

  let parse = (raw: unknown[]) => raw;
  let builder!: Knex.QueryBuilder;
  let main: Type.EntityType;

  const selects = constructor(where);

  pending.forEach(fn => fn());
  pending.clear();

  if(selects){
    builder.clearSelect();

    if(Field.is(selects)){
      builder.select({
        [selects.column]: String(selects)
      });
  
      parse = raw => raw.map(row => {
        const value = (row as any)[selects.column];
        return selects.get ? selects.get(value) : value;
      });
    }
    else if(typeof selects == "object"){
      const output = new Map<string | Field, string | number>();

      Object.getOwnPropertyNames(selects).forEach(key => {
        const value = (selects as any)[key];
    
        if(Field.is(value)){
          output.set(value, key);
          builder.select({ [key]: String(value) });
        }
      })

      parse = raw => raw.map(row => {
        const values = Object.create(selects as {});
    
        output.forEach((column, value) => {
          if(Field.is(value) && value.get)
            value = value.get((row as any)[column]);

          Object.defineProperty(values, column, { value });
        })
        
        return values as T;
      })
    }
  }

  function where(
    type: unknown,
    on?: Query.Compat | Query.Join.Function,
    join?: Query.Join.Mode
  ): any {
    if(isTypeConstructor(type)){
      if(typeof on == "string")
        throw new Error("Bad parameters.");

      return table(type, on, join);
    }

    if(Field.is(type)){
      const ref = String(type);
      const compare = (op: string) => (right: any) => {
        if(typeof right !== "number")
          right = String(right);

        builder.where(ref, op, right);
      }
  
      return {
        is: compare("="),
        isNot: compare("<>"),
        isMore: compare(">"),
        isLess: compare("<"),
        isAsc: () => builder.orderBy(ref, "asc"),
        isDesc: () => builder.orderBy(ref, "desc"),
      }
    }

    if(isFromType(type))
      return verbs(type);

    throw new Error("Invalid query.");
  }

  function verbs<T extends Type>(from: Query.FromType<T>): Query.Verbs<T> {
    const table = RelevantTable.get(from);

    if(!table)
      throw new Error(`Argument ${from} is not a query entity.`);

    return {
      delete: () => {
        builder.table(table.name).delete();
      },
      update: (data: Query.Update<any>) => {
        data = table.type.digest(data);
        builder.table(table.name).update(data);
      }
    }
  }

  function table(
    type: Type.EntityType,
    on?: Query.Compat | Query.Join.Function,
    mode?: Query.Join.Mode
  ): Query.FromType {
    let { fields, schema } = type;
    let name: string | Knex.AliasDict = type.table
    let alias: string | undefined;

    if(schema){
      alias = `$${tables.length}`;
      name = { [alias]: schema + '.' + name };
    }

    const proxy = {} as any;
    const metadata: Query.Table = { name, alias, type };

    tables.push(metadata);
    RelevantTable.set(proxy, metadata);

    fields.forEach((field, key) => {
      field = Object.create(field);
      RelevantTable.set(field, metadata);
      Object.defineProperty(proxy, key, { value: field })
    })

    if(main === undefined){
      main = type;

      const engine = type.connection?.knex || knex({
        client: "sqlite3",
        useNullAsDefault: true,
        pool: { max: 0 }
      });

      builder = engine(name).select({ count: "COUNT(*)" });
    }
    else {
      if(type.connection !== main.connection)
        throw new Error(`Joined entity ${type} does not share a connection with ${main}`);

      let callback: Knex.JoinCallback;

      switch(typeof on){
        case "function":
          callback = (table) => {
            pending.add(() => {
              on(field => {
                if (!Field.is(field))
                  throw new Error("Join assertions can only apply to fields.");

                const on = (op: string) => (right: Field) => {
                  table.on(String(field), op, String(right));
                };

                return {
                  is: on("="),
                  isNot: on("<>"),
                  isMore: on(">"),
                  isLess: on("<"),
                }
              });
            })
          }
        break;

        case "object":
          callback = (table) => {
            for (const key in on) {
              const field = type.fields.get(key);
    
              if (!field)
                throw new Error(`${key} is not a valid field in ${type}.`);
    
              const left = `${type.table}.${field.column}`;
              const right = String((on as any)[key]);
    
              table.on(left, "=", right);
            }
          }
        break;

        default:
          throw new Error(`Invalid join on: ${on}`);
      }

      switch(mode){
        case "left":
          builder.leftJoin(name, callback);
          break;
  
        case "inner":
        case undefined:
          builder.join(name, callback);
          break;
  
        case "right" as unknown:
        case "full" as unknown:
          throw new Error(`Cannot ${mode} join because that would affect ${main} which is already defined.`);
  
        default:
          throw new Error(`Invalid join type ${mode}.`);
      }
    }

    return proxy;
  }

  const query: Query = {
    then(resolve: (res: any) => any, reject: (err: any) => any){
      return builder.then(parse).then<T[]>(resolve).catch(reject);
    },
    count(){
      return builder.clone().clearSelect().count();
    },
    toString(){
      return builder.toString().replace(/```/g, "`");
    }
  }

  if(selects)
    return {
      ...query,
      async one(orFail?: boolean){
        const res = await builder.clone().limit(1).then(parse);
    
        if(res.length == 0 && orFail)
          throw new Error("Query returned no results.");
    
        return res[0] as T;
      }
    }

  return query;
}

Query.one = function one<T extends {}>(
  where: Query.Function<T>, orFail?: boolean){

  return Query(where).one(orFail);
}

function isFromType(type: any): type is Query.FromType {
  return RelevantTable.has(type);
}

export { Query }