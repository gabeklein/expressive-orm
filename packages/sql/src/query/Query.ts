import { Field, Type } from '..';
import { Syntax } from '../type/Field';
import { assign, create } from '../utils';
import { Builder as QB } from './Builder';
import { Computed } from './math';

declare namespace Query { 
  interface Table<T extends Type = any> {
    toString(): string;
    name: string;
    proxy: Query.From<T>;
    local: Map<string, Field>;
    alias?: string;
    join?: {
      as: Query.Join.Mode;
      on: Set<Syntax>;
    }
  }

  namespace Join {
    type Mode = "left" | "inner";

    // TODO: does not chain like actual Compare
    type Where = <T extends Field>(field: T) => Field.Compare<FieldOrValue<T>>;

    type Function = (on: Where) => void;

    type Equal<T extends Type = any> = { [K in keyof T]?: Field | Query.From };
    
    type On<T extends Type> = Function | Equal<T>;

    type Left<T extends Type> = Partial<From<T>>;
  }

  type Join<T extends Type> = From<T>;

  type Builder<T = any> = QB<T>;
  
  type Compare = Syntax | Compare[] | undefined;

  type From<T extends Type = Type> = {
    [K in Type.Fields<T>]: T[K] extends Field.Queries<infer U> ? U : T[K];
  }

  type Value<T = any> = T | Field<T> | Computed<T>

  type FieldOrValue<T> = T extends Value<infer U> ? U : T;

  type Where = Builder["where"] & { name: string };

  type Updates<T> = Field.Updates<T> | T;

  type Update<T extends Type> = {
    [K in Type.Fields<T>]?: Updates<T[K]>;
  }

  type Function<R> = (where: Where) => R;

  type Factory<R, A extends any[]> = (where: Where) => (...args: A) => R;
  
  type Template<A extends any[]> = (...args: A) => Query;

  type SelectTemplate<T, A extends any[]> = (...args: A) => Selects<T>;

  type Extract<T> =
    T extends Field.Returns<infer R> ? R :
    T extends From<infer U> ? { [J in Type.Fields<U>]: Extract<U[J]> } :
    T extends {} ? { [K in keyof T]: Extract<T[K]> } :
    T;

  type Asserts<T extends Field> = ReturnType<T["compare"]> & {
    asc(): void;
    desc(): void;
  };

  interface Verbs <T extends Type> {
    delete(): void;
    update(values: Query.Update<T>): void;
  }
  
  interface Selects<T> extends Query<Extract<T>[]> {
    /**
     * Returns rows which match creteria.
     */
    get(): Promise<Extract<T>[]>;
  
    /**
     * Returns the first row that matches creteria.
     * 
     * @param orFail If true, will throw an error if no results are returned.
     */
    one(orFail?: boolean): Promise<Extract<T>>;
  }
}

interface Query<T = number> extends PromiseLike<T> {
  /** Returns the SQL string generated by this query. */
  toString(): string;
}

function Query<T extends {}, A extends unknown[]>(from: Query.Factory<T, A>): Query.SelectTemplate<T, A>;

function Query<A extends unknown[]>(from: Query.Factory<void, A>): Query.Template<A>;

function Query<T extends {}>(from: Query.Function<T>): Query.Selects<T>;

/**
 * Creates a new query.
 * Will return the number of rows that would be selected or modified.
 */
function Query(from: Query.Function<void>): Query;

function Query<T = number>(factory: Query.Function<T>): Query<T> | Query.Selects<T> | Function {
  const builder = new QB(factory);
  const statement = builder.prepare();

  function runner(...inputs: any[]) {
    type Q = Query<any>;
    const params = Array.from(builder.params || [], i => inputs[i]);
    const query = create(Query.prototype) as Q; 
    
    assign(query, {
      then(resolve, reject){
        const pending = builder.selects
          ? statement.all(params)
          : statement.run(params);

        return pending.then(resolve).catch(reject);
      },
      toString(){
        return builder.toString();
      }
    } as Q);

    if (builder.selects)
      assign(query, {
        async get(){
          return statement.all(params);
        },
        async one(orFail?: boolean) {
          const res = await statement.get(params);

          if (!res && orFail)
            throw new Error("Query returned no results.");

          return res;
        }
      });

    return query;
  }

  if(builder.params){
    Object.setPrototypeOf(runner, Query.Template.prototype);
    runner.toString = () => builder.toString();
    return runner;
  }

  return runner();
}

Query.Template = function Template(){
  throw new Error("Query.Template cannot be called.");
};

Query.Template.prototype = assign(create(Query.prototype), Function.prototype)

export { Query };