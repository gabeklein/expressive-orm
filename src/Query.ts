import knex, { Knex } from 'knex';
import { format } from 'sql-formatter';

import Entity from './Entity';
import Field from './instruction/Field';
import Table from './Table';
import { escape, escapeString } from './utility';

const KNEX = knex({ client: "mysql" });

namespace Query {
  export type WhereFunction<T extends Entity> =
    (this: Where<T>, thisArg: Where<T>) => void;

  export type SelectFunction<T extends Entity, R> =
    (this: Entity.Pure<T>, thisArg: Entity.Pure<T>) => R;

  export type Options<T extends Entity, R> = {
    where?: WhereFunction<T>;
    select?: SelectFunction<T, R>;
  }

  type WhereClause<T> =
    T extends Field.Assertions<infer A> ? A : never;

  export type Where<T extends {}> =
    & { [K in Exclude<keyof T, keyof Entity>]: WhereClause<T[K]> }
    & {
      has(values: Partial<T>): void;
    }

  export type Normalize =
    (row: { [select: string]: any }, output: any) => void;

  export type Select<T extends Entity> = Entity.Pure<T>;
}

class Query<T extends Entity, S = unknown> {
  protected builder: Knex.QueryBuilder;

  public table: Table;
  public selects = new Map<string, Query.Normalize>();
  public tables = new Map<Field | undefined, string>();

  constructor(protected type: Entity.Type<T>){
    this.table = type.table;
    this.builder = KNEX.from(type.table.name);
  }

  getTableName(from?: Field){
    return this.tables.get(from) || "";
  }

  join(type: typeof Entity, on: string, foreignKey?: string){
    const foreign = type.table.name;
    const local = this.getTableName();

    // TODO: pull default from actual entity.
    const fk = escape(foreign, foreignKey || "id");
    const lk = escape(local, on);

    this.builder.joinRaw(
      `LEFT JOIN ${escape(foreign)} ON ${fk} = ${lk}`
    )

    return foreign;
  }

  toString(){
    return format(this.builder.toString());
  }

  addWhere(a: any, b: any, c: any){
    this.builder.whereRaw(`${a} ${b} ${c}`);
  }

  addSelect(name: string, callback: Query.Normalize){
    this.builder.select(name);
    this.selects.set(name, callback);
  }

  compare(
    left: Field | string,
    right: string | number | Field,
    op: string){
  
    if(typeof right === "string")
      right = escapeString(right);
    
    this.builder.whereRaw(
      `${left.toString()} ${op} ${right.toString()}`
    );
  }

  async fetch(limit?: number){
    const qs = this.builder.toString();

    if(typeof limit == "number")
      this.builder.limit(limit);

    return this.table.connection!.query(qs);
  }
  
  async get(limit?: number): Promise<S[]> {
    return this.hydrate(
      await this.fetch(limit)
    );
  }
  
  async getOne(orFail: false): Promise<S | undefined>;
  async getOne(orFail?: boolean): Promise<S>;
  async getOne(orFail?: boolean){
    const results = await this.fetch(1);

    if(results.length < 1 && orFail)
      throw new Error("No result found.");

    const output = await this.hydrate(results);

    return output[0];
  }
  
  async findOne(orFail: true): Promise<S>;
  async findOne(orFail?: boolean): Promise<S | undefined>;
  async findOne(orFail?: boolean){
    return this.getOne(orFail || false);
  }

  async hydrate(raw: any[]){
    const ops = [] as any[];
    const results = raw.map(row => {
      const output = {} as any;

      this.selects.forEach(apply => {
        const maybePromise: unknown = apply(row, output);

        if(maybePromise instanceof Promise)
          ops.push(maybePromise);
      });
      
      return output;
    });

    await Promise.all(ops);
    
    return results;
  }

  where(from: Query.WhereFunction<T>){
    const table = this.getTableName();
    const proxy = this.type.map((field) => {
      return field.where(this, table);
    });

    from.call(proxy, proxy);

    return this;
  }

  select<R>(from: Query.SelectFunction<T, R>): Query<T, R>;
  select(from: Query.SelectFunction<T, S>): this;
  select(from: Query.SelectFunction<T, any>){
    const table = this.getTableName();
    const proxy = this.type.map((field, key) => {
      return field.select(this, [key], table);
    })

    from.call(proxy, proxy);

    return this;
  }
}

export default Query;