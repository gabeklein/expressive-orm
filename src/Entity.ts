import Connection from './connection/Connection';
import Field from './Field';
import Primary from './fields/Primary';
import Query from './query/Query';

export type InstanceOf<T> = T extends { prototype: infer U } ? U : never;

const REGISTER = new Set<Entity.Type>();
const INSTRUCTION = new Map<symbol, Entity.Instruction>();

const describe = Object.getOwnPropertyDescriptor;
const define = Object.defineProperty;

declare namespace Entity {
  type Type<T extends Entity = Entity> =
    & (abstract new () => T)
    & typeof Entity;

  type List<T extends Entity> = T[];

  type DataRecusive<T> =
    T extends string ? string :
    T extends number ? number :
    T extends boolean ? boolean :
    T extends List<infer U> ? Pure<U>[] :
    T extends Entity ? Pure<T> : T;

  type Pure<T extends Entity> = {
    [K in Field<T>]-?: DataRecusive<T[K]>;
  }

  type Field<T extends Entity> = Exclude<keyof T, "table">;

  type Where<T extends Entity, R> =
    (source: Query.Type<T>, query: Query.Where) => () => R;

  type Instruction = (parent: Entity.Type, key: string) => void;
}

abstract class Entity {
  this!: Entity.Type;
  id = Primary();

  static table: string;
  static schema: string;
  static fields: Map<string, Field>;
  static deps: Set<Entity.Type>;
  static connection?: Connection;
  static focus?: { [key: string]: any };

  static ensure<T extends Entity>(
    this: Entity.Type<T>,
    connection?: Connection){

    if(!REGISTER.has(this)){
      REGISTER.add(this);

      this.table = /class (\w+?) /.exec(this.toString())![1];
      this.schema = "";
      this.connection = connection;
      this.fields = new Map();
      this.deps = new Set();
      
      const prime = new (this as any)();
      
      for(const key in prime){
        const { value } = describe(prime, key)!;
        const instruction = INSTRUCTION.get(value);    
  
        if(!instruction)
          continue;
  
        INSTRUCTION.delete(value);
        instruction(this, key);
        define(prime, key, {
          get: () => this.focus![key],
          set: is => this.focus![key] = is
        })
      }
    }

    return this;
  }

  static field(instruction: Entity.Instruction){
    const placeholder = Symbol(`ORM instruction`);
    INSTRUCTION.set(placeholder, instruction);
    return placeholder as any;
  }

  /**
   * Create an arbitary map of managed fields.
   */
  static map<T extends Entity>(
    this: Entity.Type<T>,
    getValue: (type: Field, key: Entity.Field<T>, proxy: {}) => any,
    cache?: boolean
  ){
    const proxy = {} as any;

    this.ensure();

    for(const [key, type] of this.fields)
      define(proxy, key, {
        configurable: true,
        get: () => {
          const value = getValue.call(
            proxy,
            type,
            key as Entity.Field<T>,
            proxy
          );

          if(cache !== false)
            define(proxy, key, { value });

          return value;
        }
      })
    
    return proxy;
  }

  // static select<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   from: Entity.Where<T, R>
  // ){
  //   return new Query(where => {
  //     return from(where(this), where);
  //   })
  // }

  // static insert<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   data: Insert.Values<T>
  // ){
  //   return new Insert(this, data).exec();
  // }

  // static async get<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   from: Entity.Where<T, R>
  // ){
  //   return this.select(from).get();
  // }

  // static async getOne<T extends Entity, R>(
  //   this: Entity.Type<T>,
  //   from: Entity.Where<T, R>
  // ){
  //   return this.select(from).getOne();
  // }
}

export default Entity;