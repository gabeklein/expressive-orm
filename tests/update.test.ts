import Entity, { VarChar } from "../src";
import Query from "../src/query/Query";

class Foo extends Entity {
  value = VarChar();
  color = VarChar();
}

it("will generate query", () => {
  const query = new Query(where => {
    const foo = where(Foo);

    where(foo.color).is("red");

    where.update(foo, {
      value: "Updated!",
      color: "blue"
    })
  });

  expect(query).toMatchSnapshot();
})