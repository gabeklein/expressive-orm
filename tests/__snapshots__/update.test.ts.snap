// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will generate query 1`] = `
	UPDATE
	  \`Foo\`
	SET
	  \`value\` = "Updated!",
	  \`color\` = "blue"
	WHERE
	  \`Foo\`.\`color\` = 'red'
`;
